src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java

```diff
@@ -25,39 +26,86 @@
+
+    private final int setid;
+    private int setsize;
+
     PollsetPoller(boolean read) throws IOException {
         super(read);
+        this.setsize = 0;
+        this.setid = Pollset.pollsetCreate();
     }
 
     @Override
     int fdVal() {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+        return setid;
     }
 
     @Override
-    void implRegister(int fdVal) throws IOException {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+    void implRegister(int fd) throws IOException {
+        int ret = Pollset.pollsetCtl(setid, Pollset.PS_MOD, fd,
+        Pollset.PS_POLLPRI | (this.reading() ? Net.POLLIN : Net.POLLOUT));
+        if (ret != 0) {
+            throw new IOException("Unable to register fd " + fd);
+        }
+        setsize++;
     }
 
     @Override
-    void implDeregister(int fdVal) {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+    void implDeregister(int fd) {
+        assert (setsize > 0);
+        int ret = Pollset.pollsetCtl(setid, Pollset.PS_DELETE, fd, 0);
+        assert ret == 0;
+        setsize--;
     }
 
+    /**
+      * Main poll method. The AIX Pollset library does not appear to pick up changes to the pollset
+      * (the set of fds being polled) while blocked on a call to this method. These changes happen
+      * regularly in the poll-loop thread and update thread from Poller.java.
+      * To address this difficulty, we break poll calls into 100ms sub-calls and emulate the timout.
+      */
     @Override
     int poll(int timeout) throws IOException {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+        int n;
+        switch (timeout) {
+            case 0:
+                n = pollInner(0);
+                break;
+            case Pollset.PS_NO_TIMEOUT:
+                do { n = pollInner(100); } while (n == 0);
+                break;
+            default:
+                Instant end = Instant.now().plusMillis(timeout);
+                do { n = pollInner(100); } while (n == 0 && Instant.now().isBefore(end));
+                break;
+        }
+        return n;
+    }
+
+    int pollInner(int subInterval) throws IOException {
+        // The poll loop may start polling before any fds have been registered. But, if we use set
+        // size 0 to allocatePollArray, it will return the null address and pollsetPoll will complain.
+        // We avoid that by just passing set size 1, and letting poll block for subInterval.
+        long buffer = Pollset.allocatePollArray(setsize > 0 ? setsize : 1);
+        int n = Pollset.pollsetPoll(setid, buffer, setsize, subInterval);
+        for (int i=0; i<n; i++) {
+            long eventAddress = Pollset.getEvent(buffer, i);
+            int fd = Pollset.getDescriptor(eventAddress);
+            polled(fd);
+        }
+        Pollset.freePollArray(buffer);
```

Comment:
If I read this correctly, there is a malloc/free per poll, is that right?

Reply:
If you look at the Poller implementations on the other platform you'll see that they allocate the poll array at construction time, I think that is what you want here too.

Reply:
You are correct that there is a malloc/free per poll. I made this decision intentionally.

The other implementations allocate a fixed size array at object creation. This imposes an upper limit on the number of fds that can be polled. I made the decision to move this allocation to the poll call and track the size of the poll set. This way, (1) the only limit to the size of the poll set is the system limit on number of open fds (if you can open an fd it can be polled) (2) there is no wasted space to an arbitrarily large array when there are only a few fds polled.

Reply:
there will be some performance impact  in "allocation/de-allocation"  memory  per poll approach.

Reply:
> The other implementations allocate a fixed size array at object creation. This imposes an upper limit on the number of fds that can be polled. I made the decision to move this allocation to the poll call and track the size of the poll set. This way, (1) the only limit to the size of the poll set is the system limit on number of open fds (if you can open an fd it can be polled) (2) there is no wasted space to an arbitrarily large array when there are only a few fds polled.

I don't have any experience on AIX but docs page for pollset_poll suggests it populates the array with the file descriptors that have pending events. If I read this correctly, then it's nothing to do with the number of file descriptors registered and it works similarly to /dev/poll, epoll and kqueue. So I think maybe look at this again.

Reply:
In regards to the speed and fragmentation issues. I think these are compelling reasons to consider the other approach (Option 2 below).

With respect to memory being returned to the OS: I would expect this issue to be minimal because the memory would 'stick' to the process and just be recycled.

> If I read this correctly, then it's nothing to do with the number of file descriptors registered

The implementation as it is written keeps track of the number of fds registered and only allocates memory for that number (0-2 in my testing) when the poll starts. The other implementations allocate space for 512 pollfds at object creation. You are correct that the library only _populates_ the space if there is an event, but space has to be allocated before hand.

The memory requirements per poller are actually doubled because there is always a reader and writer created.

### Option 1: Memory allocated per-poll

Pros:
- Memory footprint is minimal ~ `1 * sizeof(struct pollfd)`.

Cons:
- Possible performance issues (speed).
- Possible memory fragmentation issues.

### Option 2: Memory allocated on Poller creation

Pros:
- No issues with memory fragmentation or allocation performance (since malloc/free calls are minimized).

Cons:
- Arbitrary upper limit on number of events per poll.
- Much larger memory footprint `512 * sizeof(struct pollfd)`

---

Option 2 seems to be the popular choice, and I'm amenable to changing that. It does seems a bit silly to allocate 2 * 512 * 8 bytes == 8 KiB for each Poller when I've only seen 4 * 8 bytes in use.

Reply:
Are there alternatives like using alloca?

Reply:
> Are there alternatives like using alloca?

Cool suggestion! I actually hadn't heard of it before. I am still not totally sure how it's different from just using local variables which should be on the stack anyway. From a quick man-page search it looks viable to do this here.

Another possibility: I could use Option 2, but not use such a large value for MAX_FDS. The max we choose just dictates how many results are returned by each call to pollset_poll. Any fds not returned in one call should just be returned the next call, and the poll loop is constantly making poll calls.

Some quick testing suggests that pollset does a pretty good job of returning unseen events first. Polling a set of 8 files that could all be read with a pollset size of 1, I saw 8 unique fds.

So, I'm leaning towards Option 2 with MAX_FDS of ~16. Does that seems like a reasonable compromise?

Reply:
> Option 2 seems to be the popular choice, and I'm amenable to changing that. It does seems a bit silly to allocate 2 * 512 * 8 bytes == 8 KiB for each Poller when I've only seen 4 * 8 bytes in use.

There are 2 Poller threads per VM. They run in a loop that does a blocking poll (pollset_poll in the discussion here) to pick up events in bulk. 8K doesn't seem too bad and you could reduce the 512 to a small batch size if you want.

Another point is that there is configuration knob to control the registration. In "direct" mode, virtual threads that need to block on I/O will will attempt to directly arm the file descriptor. This means that implRegister will be called concurrently by many threads, the implRegister implementation in the changes here will increment and decrement setsize without synchronization. You may not have seen this as the default uses "indirect" mode.

Reply:
I agree. 8K per Poller threads sounds still affordable.

Reply:
Ok. I have changed the implementation to be similar to that of the other platforms--including the size of the buffer.